### PR TITLE
redirect to paypal after successful donation

### PR DESCRIPTION
**Pull request by JeroenDeDauw**

### CODE REVIEW

File: src/Domain/PayPalUrlGenerator.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ */
+class PayPalUrlGenerator {
+
+	private $config;
+
+	public function __construct( PayPalConfig $config ) {
+		$this->config = $config;
+	}
+
+	public function generateUrl( int $id, float $amount, int $interval, string $accessToken, string $updateToken ) {
```
Comment (JeroenDeDauw) [inline src/Domain/PayPalUrlGenerator.php]: I had to check the caller to verify this is the donation id, so $donationId is better

File: src/Domain/PayPalUrlGenerator.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	private function getSubscriptionParams( $amount, $interval ): array {
+		return [
+			'cmd' => '_xclick-subscriptions',
+			'no_shipping' => '1',
+			'src' => '1', # recurring
+			'sra' => '1', # reattempt
+			'srt' => '0', # number of payment cycles (0 for infinite)
+			'a3' => $amount,
+			'p3' => $interval,
+			't3' => 'M', # monthly payment cycle
```
Comment (JeroenDeDauw) [inline src/Domain/PayPalUrlGenerator.php]: How about using constants for these instead of adding comments?

File: src/Factories/FunFunFactory.php [hunk 3]
```diff
@@ -664,6 +666,14 @@ private function newAddDonationMailer(): TemplateBasedMailer {
 		);
 	}
 
+	public function newPayPalUrlGenerator() {
+		return new PayPalUrlGenerator( $this->getPayPalConfig() );
+	}
+
+	private function getPayPalConfig() {
+		return ( PayPalConfig::newFromConfig( $this->config['paypal'] ) );
```
Comment (JeroenDeDauw) [inline src/Factories/FunFunFactory.php]: What's with the brackets?

File: src/Domain/PayPalUrlGenerator.php (hunk 2 above)
Comment (KaiNissen) [inline src/Domain/PayPalUrlGenerator.php]: Good point, changed that.

File: src/Domain/PayPalUrlGenerator.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+	private function getIntervalDependentParameters( float $amount, int $interval ): array {
+		if ( $interval > 0 ) {
+			return $this->getSubscriptionParams( $amount, $interval );
+		}
+
+		return $this->getSinglePaymentParams( $amount );
+	}
+
+	private function getSubscriptionParams( float $amount, int $interval ): array {
```
Comment (gbirke) [inline src/Domain/PayPalUrlGenerator.php]: Maybe add a function comment, and add a link to the paypal documentation  where these cryptic parameter names are explained?

**Comment by gbirke** [comment]
+1

File: src/Domain/PayPalConfig.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+		$this->payPalAccountAddress = $payPalAccountAddress;
+		$this->payPalBaseUrl = $payPalBaseUrl;
+		$this->notifyUrl = $notifyUrl;
+		$this->returnUrl = $returnUrl;
+		$this->cancelUrl = $cancelUrl;
+		$this->itemName = $itemName;
+	}
+
+	/**
+	 * @param String[] $config
```
Comment (JeroenDeDauw) [inline src/Domain/PayPalConfig.php]: string[]

File: src/Domain/PayPalConfig.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	 * @return PayPalConfig
+	 */
+	public static function newFromConfig( array $config ) {
```
Comment (JeroenDeDauw) [inline src/Domain/PayPalConfig.php]: If we add constants for these keys, then we avoid a class of errors and make navigation and refactoring easier.

**Pull request merged**